### DISCUSSION TITLE
Which algorithm does AIR use to encrypt SQL database?

### DISCUSSION THREAD

**Discussion by DavidWithTuxedo** [airsdk/Adobe-Runtime-Support#1921] [Q&A] [answered]
Hello, dear developers!

I have a question, which standard does the encryption of SQL database files in AIR comply with? [Using encryption with SQL databases](https://help.adobe.com/en_US/air/html/dev/WS8AFC5E35-DC79-4082-9AD4-DE1A2B41DAAF.html) Although this page claims it's the "AES-CBC-MAC", all I get with the "EncryptionKeyGenerator" is a string of 16-bit ciphers instead of 128/256-bit.

For testing if it meets the standard, I tried to use tools like the "openssl", but can't decrypt my encrypted "test.dat". Is it really AES-CBC-MAC? And I notice the usual way to call it in ActionScript is like this,
```java
flash.data.SQLConnection.open(dbFile, air.SQLMode.CREATE, null, false, 1024, encryptionKey);
```
Is there any doc on the exact process of this function?

Thanks in advance!

**Comment by ajwfrost** [comment, marked as answer]
Hi

I've not looked at that `EncryptionKeyGenerator` code (although it's open-source I believe) but according to the `SQLConnection` documentation, the key should be a 16-byte ByteArray (so 128-bit key).  There's actually a check within `SQLConnection` for the length of the byte array..

Under the hood, AIR is using the SQLite encryption extension, with the AES-128 algorithm in CCM mode. See https://www.sqlite.org/see/doc/release/www/readme.wiki

From what I've seen, each page of the database file is encrypted separately which means it wouldn't just be possible to treat the database file as if it's an encrypted file, and try to decrypt it in one go.

thanks
